src/main/kotlin/com/openlattice/indexing/BackgroundExpiredDataDeletionService.kt

```diff
@@ -0,0 +1,250 @@
+    private val expirationLocks: IMap<UUID, Long> = hazelcastInstance.getMap(HazelcastMap.EXPIRATION_LOCKS.name)
+
+    //necessary??????
+    init {
+        expirationLocks.addIndex(QueryConstants.THIS_ATTRIBUTE_NAME.value(), true)
+    }
+
+    private val taskLock = ReentrantLock()
+
+    @Suppress("UNCHECKED_CAST", "UNUSED")
+    @Scheduled(fixedRate = MAX_DURATION_MILLIS)
+    fun scavengeIndexingLocks() {
+        expirationLocks.removeAll(
+                Predicates.lessThan(
+                        QueryConstants.THIS_ATTRIBUTE_NAME.value(),
+                        System.currentTimeMillis()
+                ) as Predicate<UUID, Long>
+        )
+    }
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = DATA_DELETION_RATE)
+    fun deleteExpiredDataFromEntitySets() {
+        logger.info("Starting background expired data deletion task.")
+        //Keep number of expired data deletion jobs under control
+        if (taskLock.tryLock()) {
+            try {
+                if (indexerConfiguration.backgroundExpiredDataDeletionEnabled) {
+                    val w = Stopwatch.createStarted()
+                    //We shuffle entity sets to make sure we have a chance to work share and index everything
+                    //lock entity sets we are working on
+                    val lockedEntitySets = entitySets.values
+                            .shuffled()
+                            .filter { tryLockEntitySet(it) } //filters out entitysets that are already locked, and the entitysets we're working on are now locked in the IMap
+                            .filter { it.name != "OpenLattice Audit Entity Set" } //TODO: Clean out audit entity set from prod
+
+                    //delete expired data
+                    val totalDeleted = lockedEntitySets
+                            .parallelStream()
+                            .filter { !it.isLinking }
+                            .mapToInt { deleteExpiredData(it) }
+                            .sum()
+
+                    //unlock the entitysets we were working on
+                    lockedEntitySets.forEach(this::deleteIndexingLock)
+
+                    logger.info(
+                            "Completed deleting {} expired elements in {} ms.",
+                            totalDeleted,
+                            w.elapsed(TimeUnit.MILLISECONDS)
+                    )
+                } else {
+                    logger.info("Skipping expired data deletion as it is not enabled.")
+                }
+            } finally {
+                taskLock.unlock()
+            }
+        } else {
+            logger.info("Not starting new expired data deletion job as an existing one is running.")
+        }
+    }
+
+    //deletes expired data from data and ids table in postgres and elasticsearch
+    private fun deleteExpiredData(entitySet: EntitySet): Int {
+        logger.info(
```

Comment:
I would also move this log line after the check whether there's an expiration policy to avoid log spam (you could also add that check as a filter when you're loading entity sets in the fixed rate task)

Reply:
omg yes will do